File: 255-verify-preorder-sequence-in-binary-search-tree/255-verify-preorder-sequence-in-binary-search-tree.cpp
class Solution {
public:
    bool verifyPreorder(vector<int>& preorder) 
    {
        //In BST only one dip allowed
        //If two dips, then return false
        
        int check = INT_MIN;
        
        stack <int> st;
        
        for(auto c : preorder)
        {
            while(!st.empty() && st.top() < c)
            {
                check = st.top();
                st.pop();
            }
            if(check != INT_MIN && check > c)
                return false;
            st.push(c);
        }
        
        return true;
        
    }
};

File: 255-verify-preorder-sequence-in-binary-search-tree/255-verify-preorder-sequence-in-binary-search-tree.py
class Solution:
    def verifyPreorder(self, preorder: List[int]) -> bool:
        
        check, stack = None, []
        for num in preorder:
            while stack and num > stack[-1]:
                check = stack.pop()
            if check!=None and check > num:
                return False
            stack.append(num)
        return True